packages/tiktok-video-element/tiktok-video-element.js

```diff
@@ -0,0 +1,252 @@
+      return o;
+    }, {});
+  }
+
+  render() {
+    const id = this.getAttribute('video-id');
+    if (!id) return;
+
+    this.loadComplete = this._createPublicPromise();
+
+    const params = {
+      controls: this.attrs.controls === '' ? null : 0,
+      autoplay: this.attrs.autoplay,
+      loop: this.attrs.loop,
+      mute: this.attrs.muted,
+      playsinline: this.attrs.playsinline,
+      preload: this.attrs.preload ?? 'metadata',
+      enablejsapi: 1,
+      showinfo: 0,
+      rel: 0,
+      iv_load_policy: 3,
+      modestbranding: 1,
+      ...this.config,
+    };
+    const src = `${TikTokVideoElement.EMBED_BASE}/${id}?${this.serialize(params)}`;
+
+    this.shadowRoot.innerHTML = `
+      <style>
+        :host { display:inline-block;position:relative;width:100%;height:100%; }
+        iframe { position:absolute;top:0;left:0;width:100%;height:100%;border:0;}
+      </style>
+      <iframe src="${src}" allow="autoplay; fullscreen" title="TikTok video"></iframe>
+    `;
+
+    this.iframe = this.shadowRoot.querySelector('iframe');
+  }
+
+  boolToBinary(props) {
+    let p = {};
+    for (let key in props) {
+      let val = props[key];
+      if (val === true || val === '') p[key] = 1;
+      else if (val === false) p[key] = 0;
+      else if (val != null) p[key] = val;
+    }
+    return p;
+  }
+
+  serialize(props) {
+    return String(new URLSearchParams(this.boolToBinary(props)));
+  }
+
+  handleMessage(event) {
+    const msg = event.data;
+    if (!msg?.['x-tiktok-player']) return;
+
+    switch (msg.type) {
+      case 'onPlayerReady':
+        this.loadComplete.resolve();
+        this.dispatchEvent(new Event('ready'));
+        break;
+
+      case 'onStateChange':
+        this.playerState = msg.value;
+        this._paused = msg.value !== TikTokVideoElement.PlayerState.PLAYING;
+        this.dispatchEvent(new CustomEvent('statechange', { detail: msg.value }));
+        this.dispatchEvent(
+          new Event(
+            msg.value === TikTokVideoElement.PlayerState.PLAYING
+              ? 'play'
+              : msg.value === TikTokVideoElement.PlayerState.PAUSED
+              ? 'pause'
+              : msg.value === TikTokVideoElement.PlayerState.ENDED
+              ? 'ended'
+              : ''
+          )
+        );
+        break;
+
+      case 'onCurrentTime':
+        this._currentTime = msg.value.currentTime;
+        this._duration = msg.value.duration;
+        this.dispatchEvent(new CustomEvent('durationchange', { detail: this._duration }));
+        this.dispatchEvent(new CustomEvent('timeupdate', { detail: msg.value }));
+        break;
+
+      case 'onVolumeChange':
+        this._volume = msg.value;
+        this.dispatchEvent(new CustomEvent('volumechange', { detail: msg.value }));
+        if (msg.value === 0) {
+          this._muted = true;
+        } else {
+          this._muted = false;
+        }
+        this.dispatchEvent(new CustomEvent('mutechange', { detail: this._muted }));
```

Comment:
all these events should use the `Event` class without detail and they have to be native htmlvideoelement events.

`ready`, `statechange`, `mutechange` don't exist...